sycl/include/sycl/ext/intel/esimd/common.hpp

```diff
@@ -352,6 +350,239 @@ template <__ESIMD_NS::atomic_op Op> constexpr int get_num_args() {
+            DS == lsc_data_size::u16u32 || DS == lsc_data_size::u16u32h)) ||
+          (sizeof(T) == 8 && DS == lsc_data_size::u64),
+      "Data type does not match data size");
+}
+
+template <typename T, lsc_data_size DS>
+constexpr lsc_data_size finalize_data_size() {
+  check_lsc_data_size<T, DS>();
+  if (DS != lsc_data_size::default_size)
+    return DS;
+  else if (sizeof(T) == 1)
+    return lsc_data_size::u8;
+  else if (sizeof(T) == 2)
+    return lsc_data_size::u16;
+  else if (sizeof(T) == 4)
+    return lsc_data_size::u32;
+  else if (sizeof(T) == 8)
+    return lsc_data_size::u64;
+  else
+    return DS;
+}
+
+enum class lsc_vector_size : uint8_t {
+  n1 = 1,
+  n2 = 2,
+  n3 = 3,
+  n4 = 4,
+  n8 = 5,
+  n16 = 6,
+  n32 = 7,
+  n64 = 8,
+};
+
+template <int VS> constexpr void check_lsc_vector_size() {
+  static_assert(VS == 1 || VS == 2 || VS == 3 || VS == 4 || VS == 8 ||
+                    VS == 16 || VS == 32 || VS == 64,
+                "Unsupported vector size");
+}
+
+template <lsc_vector_size VS> constexpr void check_lsc_vector_size() {
+  static_assert(VS == lsc_vector_size::n1 || VS == lsc_vector_size::n2 ||
+                    VS == lsc_vector_size::n3 || VS == lsc_vector_size::n4 ||
+                    VS == lsc_vector_size::n8 || VS == lsc_vector_size::n16 ||
+                    VS == lsc_vector_size::n64 || VS == lsc_vector_size::n32,
+                "Unsupported vector size");
+}
+
+template <lsc_vector_size VS> constexpr uint8_t to_int() {
+  check_lsc_vector_size<VS>();
+  switch (VS) {
+  case lsc_vector_size::n1:
+    return 1;
+  case lsc_vector_size::n2:
+    return 2;
+  case lsc_vector_size::n3:
+    return 3;
+  case lsc_vector_size::n4:
+    return 4;
+  case lsc_vector_size::n8:
+    return 8;
+  case lsc_vector_size::n16:
+    return 16;
+  case lsc_vector_size::n32:
+    return 32;
+  case lsc_vector_size::n64:
+    return 64;
+  default:
+    return 1;
+  }
+}
+
+template <int VS> constexpr lsc_vector_size to_lsc_vector_size() {
+  check_lsc_vector_size<VS>();
+  switch (VS) {
+  case 1:
+    return lsc_vector_size::n1;
+  case 2:
+    return lsc_vector_size::n2;
+  case 3:
+    return lsc_vector_size::n3;
+  case 4:
+    return lsc_vector_size::n4;
+  case 8:
+    return lsc_vector_size::n8;
+  case 16:
+    return lsc_vector_size::n16;
+  case 32:
+    return lsc_vector_size::n32;
+  case 64:
+    return lsc_vector_size::n64;
+  default:
+    return lsc_vector_size::n1;
+  }
+}
+
+enum class lsc_data_order : uint8_t {
+  nontranspose = 1,
+  transpose = 2,
+};
+
+template <cache_hint Hint> class cache_hint_wrap {
+  template <cache_hint...> struct is_one_of_t;
+  template <cache_hint Last>
+  struct is_one_of_t<Last>
+      : std::conditional_t<Last == Hint, std::true_type, std::false_type> {};
+  template <cache_hint Head, cache_hint... Tail>
+  struct is_one_of_t<Head, Tail...>
+      : std::conditional_t<Head == Hint, std::true_type, is_one_of_t<Tail...>> {
+  };
```

Comment:
nit, can easily be ignored.

I'd prefer this to be written such that it could look closer to https://godbolt.org/z/zrT3Ya6qx, i.e., using `constexpr` function instead of metaprogramming.

Reply:
Thank you.
Here I only copied it from experimental. We have several/many like this. It deserves a special/separate PR.